help_function/src/onedpl_test_transform_output_iterator.cpp

```diff
@@ -0,0 +1,441 @@
+    auto sycl_result_begin = sycl::malloc_shared<int>(buffer_size, q);
+
+    // 3. run algorithms
+    auto transformation1 = [](float item) { return (int)(item * 2.0f); };
+    auto transformation2 = [](double item) { return (float)(item + 1.0); };
+
+    double init = 0.5;
+
+    ::std::fill_n(sycl_source_begin, buffer_size, init);
+
+    auto tr1_host_result_begin = oneapi::dpl::make_transform_output_iterator(
+        sycl_result_begin, transformation1);
+    auto tr2_host_result_begin = oneapi::dpl::make_transform_output_iterator(
+        tr1_host_result_begin, transformation2);
+
+    ::std::vector<int> expected_res(buffer_size,
+                                    (int)((float)(init + 1.0) * 2.0f));
+
+    return test_copy(oneapi::dpl::execution::make_device_policy(q),
+                     sycl_source_begin, sycl_source_begin + buffer_size,
+                     tr2_host_result_begin, sycl_result_begin, buffer_size,
+                     expected_res.begin());
+  }
+};
+
+struct test_zip_iterator {
+  int operator()(size_t buffer_size) {
+
+    sycl::queue q{};
+
+    auto sycl_source_begin = sycl::malloc_shared<float>(buffer_size, q);
+    auto sycl_result_begin1 = sycl::malloc_shared<float>(buffer_size, q);
+    auto sycl_result_begin2 = sycl::malloc_shared<float>(buffer_size, q);
+
+    auto zip =
+        oneapi::dpl::make_zip_iterator(sycl_result_begin1, sycl_result_begin2);
+
+    auto transformation1 = [](const auto &item) {
+      return ::std::make_tuple(item - 1.0f, item * item);
+    };
+    auto tr1_host_result_begin =
+        dpct::make_transform_output_iterator(zip, transformation1);
+
+    float init = 2.0f;
+
+    ::std::fill_n(sycl_source_begin, buffer_size, init);
+
+    ::std::vector<float> expected_res1(buffer_size, init - 1.0f);
+    ::std::vector<float> expected_res2(buffer_size, init * init);
+    auto zip_res = oneapi::dpl::make_zip_iterator(expected_res1.begin(),
+                                                  expected_res2.begin());
+
+    ::std::copy(oneapi::dpl::execution::make_device_policy(q),
+                sycl_source_begin, sycl_source_begin + buffer_size,
+                tr1_host_result_begin);
+
+    int failed_tests = 0;
+    failed_tests +=
+        ASSERT_EQUAL_N(expected_res1.begin(), sycl_result_begin1, buffer_size,
+                       "Wrong result from copy with transform_output_iterator");
+    failed_tests +=
+        ASSERT_EQUAL_N(expected_res2.begin(), sycl_result_begin2, buffer_size,
+                       "Wrong result from copy with transform_output_iterator");
+
+    return failed_tests;
+  }
+};
+
+template <typename OutputIterator1, typename UnaryFunc>
+auto attempt_to_dangle(OutputIterator1 out1, UnaryFunc unary) {
+  auto toi = oneapi::dpl::make_transform_output_iterator(out1, unary);
+  // Leave scope of the transform output iterator, then assign data
+  // to the wrapper after the fact.
+  return *toi;
```

Comment:
Just as clarification, do we intend to the behavior to be that the transformation is still performed? I see the expected result is filled with `init * init`.